avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/FeatureListGeneration.scala

```diff
@@ -160,46 +194,120 @@ class FeatureListGeneration {
-              })
-          inputStream.close()
-        })
-
-    if (writer != null) writer.close()
-    fileSystem.delete(new Path(tmpFeatureListPath), ENABLE_RECURSIVE)
+    // first get a set of tensor names for which temporary feature list files have been collected
+    val allColsToWriteFeatureLists = new mutable.HashSet[String]()
+    val tmpFeatureListDir = s"${params.workingDir.rootPath}/$TMP_FEATURE_LIST"
+    allColsToWriteFeatureLists ++= fileSystem.listStatus(new Path(tmpFeatureListDir)).filter(_.isDirectory())
+      .filter(_.getPath.getName.startsWith(s"$COLUMN_NAME=")).map(_.getPath.getName.split(s"$COLUMN_NAME=").last)
+    val tensorGroups = new mutable.ArrayBuffer[Array[String]]
+    params.tensorsSharingFeatureLists match {
+      case Some(tensorSharingGroups) => {
+        tensorGroups ++= tensorSharingGroups
+        allColsToWriteFeatureLists --= tensorSharingGroups.flatten.toSet
+      }
+      case None => ()
+    }
+    if (allColsToWriteFeatureLists.nonEmpty) {
+      tensorGroups ++= allColsToWriteFeatureLists.map(tensor => Array(tensor))
+    }
+    // merge and write feature lists for output tensors with shared feature list setting
+    tensorGroups.foreach( // each element is an array containing the output tensor names sharing one feature list
+      tensors => {
+        // for the current group of tensors sharing one feature list, accumulate feature entry count in a hashmap
+        val featureEntriesWCount = new mutable.HashMap[String, Int]().withDefaultValue(0)
+        // get a list of tensors where a prefix needs to be removed when accumulating feature entry count and added back
+        // when writing the feature entries out. These are ntv tensors with feature sharing setting
+        val tensorsWithPrefix = new mutable.HashMap[String, String]()
+        tensors.foreach( // go over the temporary feature list files for each output tensor
+          tensor => {
+            // determine whether the tensor needs a special prefix treatment
+            val needProcessPrefix: Boolean = tensors.size > 1 && ntvTensors.contains(tensor)
+            val prefix = new mutable.HashSet[String]() // to make sure there is only one prefix per tensor
+            val featureListDirForCurrentTensor = new Path(s"$tmpFeatureListDir/$COLUMN_NAME=$tensor")
+            val filesIterator = fileSystem.listFiles(featureListDirForCurrentTensor, ENABLE_RECURSIVE)
+            while (filesIterator.hasNext) {
+              val tmpFeatureListPath = filesIterator.next().getPath
+              val fileInputStream = fileSystem.open(tmpFeatureListPath)
+              scala.io.Source.fromInputStream(fileInputStream, UTF_8.name()).getLines()
+                .foreach(
+                  line => {
+                    // the format of each line is feature_entry,count
+                    val featureEntry = if (needProcessPrefix) {
+                      val prefixCurrentLine = line.split(SPLIT_REGEX).head.split(',').head
```

Comment:
can we do a this split on `SPLIT_REGEX ` one time, and reuse the result downstream.

Reply:
Good point. Will address.